section-09-differential-testing/model-package/differential_tests/test_differential.py

```diff
@@ -0,0 +1,36 @@
+import math
+from pathlib import Path
+import os
+
+
+ACCEPTABLE_MODEL_DIFFERENCE = float(os.environ.get('ACCEPTABLE_MODEL_DIFFERENCE'))
+
+
+def test_difference_between_current_and_candidate_models(sample_input_data):
```

Comment:
As I go through the code I understand that we have the following tests:

**test_differencial** (this file), where we supposedly already have a model in production and we want our model to perform similarly.

**test_model_accuracy**: that I don't understand what is tested and I added a comment there.

**test_prediction.py**: where we test one prediction.

**Differential testing - my thoughts**

We are making our students deploy a model for the first time, so comparing the model's predictions with a model that exists, might be confusing. So we need to introduce first differential testing, introduce when it is used, and potential advantages and shortcomings. With differential testing, if we are comparing with a model that exists already, and we create a new model, that is way better, in principle, I do not expect the predictions to be within a certain tolerance, i expect them to be way better.

**test_prediction - my thoughts:**
In the research environment section (section 4) and in section where I talk about reproducibility (section 3 I think), I stress the fact that the model in production should be identical (or as similar as possible) to the one we created in the research environment. And I promise students that we will see how to test this later on in the course (this section). 

So I think that when introducing this section, it would be good to first re-cap shortly on reproducibility.

In cell 13 in [this notebook](https://github.com/trainindata/deploying-machine-learning-models/blob/master/section-04-research-and-development/08-final-machine-learning-pipeline.ipynb), I stored the pipeline. We could bring that notebook here to create and store a set of predictions first (remind students that this notebook happened, so it helps them visualize how things come together). Once we save the predictions from the notebook, those would be the ones we have in the research environment, and then we compare the predictions in production with those. This way, we are not comparing just one prediction as per your script later, but a whole bunch. Which will allow us to test the model accuracy also in production, which at the moment I don't think we are doing. And, we bring everything together this way.

Reply:
See notebook updates and new tests. Basically we'll have to groups of tests:

1. "New Model" and/or "significant model update" test: `test_compare_against_research_environment_predictions` which uses the `research_model_test_results.csv` generated from the updated notebook in the research-notebook directory
2. "Business as usual" tests: `test_difference_between_current_and_candidate_models` which checks for model drift from one version to another (best for catching errors in config/feature engineering)